### PR TITLE
TSE Code Review: Ruby - My Enumerables

### PR DESCRIPTION
**Pull request by aecordova**
# [Ruby] Project 2 - My Enumerable

In this project we've done the following:
Replicate the functionality of some of the most used methods in the ruby enumerable module: 
- each 
- each_with_index 
- select 
- all?
- none?
- any? 
- count 
- map
- inject

### CODE REVIEW

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice job :+1: 
I appreciate the effort you have put into this project. I tested your code and I can say most have passed against the edge cases but we need to fix the failing ones.
Check my comments for suggestions.

### Martin Afani

File: my_enumerable_module.rb [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+# Extending Enumerable module with replicated methods for the following:
+# each, each_with_index, select, all?, none?, any?, count, map and inject
+#
+module Enumerable
+  def my_each
+    arr = self # this is to prevent a redundant-use-of-self error
+    0.upto(arr.length - 1) do |i|
+      yield(arr[i])
+    end
+  end
```
Comment (whiz25) [inline my_enumerable_module.rb]: - [x] should be identical and returns the same thing as ruby's [`each`](https://ruby-doc.org/core-2.6.4/Array.html#method-i-each). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns an Enumerator if no block is given

File: my_enumerable_module.rb [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+
+  def my_each_with_index
+    arr = self # this is to prevent a redundant-use-of-self error
+    0.upto(arr.length - 1) do |i|
+      yield(arr[i], i)
+    end
+  end
```
Comment (whiz25) [inline my_enumerable_module.rb]: - [x] should be identical and returns the same thing as ruby's [`each_with_index`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-each_with_index). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_each_with_index.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns an enumerator if no block is given

File: my_enumerable_module.rb [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+  def my_select
+    arr = self # this is to prevent a redundant-use-of-self error
+    result = []
+    0.upto(arr.length - 1) do |i|
+      result << arr[i] if yield(arr[i])
+    end
+    result
+  end
```
Comment (whiz25) [inline my_enumerable_module.rb]: - [x] should be identical and returns the same thing as ruby's [`select`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-select). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_select.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns an enumerator if no block is given

File: my_enumerable_module.rb [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+  def my_all?
+    arr = self # this is to prevent a redundant-use-of-self error
+    response = true
+    0.upto(arr.length - 1) do |i|
+      response = false unless yield(arr[i])
+    end
+    response
+  end
```
Comment (whiz25) [inline my_enumerable_module.rb]: - [x] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] when no block or argument is given returns true when none of the collection members are false or nil
- [x] when a class is passed as an argument returns true if all of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern

File: my_enumerable_module.rb [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+  def my_any?
+    arr = self # this is to prevent a redundant-use-of-self error
+    response = false
+    0.upto(arr.length - 1) do |i|
+      response = true if yield(arr[i])
+    end
+    response
+  end
```
Comment (whiz25) [inline my_enumerable_module.rb]: - [x] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true if at least one of the collection is not false or nil
- [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [x] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: my_enumerable_module.rb [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+  def my_none?
+    arr = self # this is to prevent a redundant-use-of-self error
+    response = true
+    0.upto(arr.length - 1) do |i|
+      response = false if yield(arr[i])
+    end
+    response
+  end
```
Comment (whiz25) [inline my_enumerable_module.rb]: - [x] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true only if none of the collection members is true
- [x] when a class is passed as an argument returns true if none of the collection is a member of such class
- [x] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: my_enumerable_module.rb [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+  def my_count(single_value = nil)
+    arr = self
+    counter = 0
+    if single_value
+      arr.my_each { |val| counter += 1 if val == single_value }
+    else
+      arr.my_each { |val| counter += 1 if yield(val) }
+    end
+    counter
+  end
```
Comment (whiz25) [inline my_enumerable_module.rb]: - [x] should be identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns the number of items in enum through enumeration

File: my_enumerable_module.rb [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+
+  def my_map(a_proc = nil)
+    arr = self # this is to prevent a redundant-use-of-self error
+    if block_given? && a_proc.nil?
+      arr.my_each_with_index { |v, i| arr[i] = yield(v) }
+    else
+      arr.my_each_with_index { |v, i| arr[i] = a_proc.call(v) }
+    end
+    arr
+  end
```
Comment (whiz25) [inline my_enumerable_module.rb]: - [x] should be identical and returns the same thing as ruby's [`map`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-map). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_map.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] returns a new array with the results of running block once for every element in enum.
- [x] returns an Enumerator if no block is given

File: my_enumerable_module.rb [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+    accum = obj[init]
+    if block_given?
+      (init + 1).upto(obj.length - 1) { |i| accum = yield(accum, obj[i]) }
+    else
+      (init + 1).upto(obj.length - 1) { |i| accum = accum.send(oper, obj[i]) }
+    end
+    accum
+  end
+  # rubocop:enable Metrics/AbcSize, Metrics/MethodLength
+end
```
Comment (whiz25) [inline my_enumerable_module.rb]: - [x] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] combines all elements of enum by applying a binary operation, specified by a block
- [x] when a symbol is specified combines each element of the collection by applying the symbol as a named method

File: my_enumerable_module.rb (hunk 1 above)
Comment (aecordova) [reply my_enumerable_module.rb]: Corrected

File: my_enumerable_module.rb (hunk 2 above)
Comment (aecordova) [reply my_enumerable_module.rb]: corrected

File: my_enumerable_module.rb (hunk 4 above)
Comment (aecordova) [reply my_enumerable_module.rb]: corrected

File: my_enumerable_module.rb (hunk 3 above)
Comment (aecordova) [reply my_enumerable_module.rb]: corrected

File: my_enumerable_module.rb (hunk 5 above)
Comment (aecordova) [reply my_enumerable_module.rb]: corrected

File: my_enumerable_module.rb (hunk 6 above)
Comment (aecordova) [reply my_enumerable_module.rb]: Corrected

File: my_enumerable_module.rb (hunk 7 above)
Comment (aecordova) [reply my_enumerable_module.rb]: Corrected

File: my_enumerable_module.rb (hunk 8 above)
Comment (aecordova) [reply my_enumerable_module.rb]: Corrected

File: my_enumerable_module.rb (hunk 9 above)
Comment (aecordova) [reply my_enumerable_module.rb]: Corrected

**Review by Forison — APPROVED** [review]
### Status: Approved 💚 
# Hi @aecordova,
Good job on implementing the changes suggested by my colleague earlier.
Go ahead and merge the project and proceed to the next project 
### Happy coding
### Addo Forison

**Pull request merged**